src/docker/custom.rs

```diff
@@ -138,3 +134,86 @@ impl<'a> Dockerfile<'a> {
         }
     }
 }
+
+fn docker_package_name(metadata: &CargoMetadata) -> String {
+    // a valid image name consists of the following:
+    // - lowercase ASCII letters
+    // - digits
+    // - a period
+    // - 1-2 underscores
+    // - 1 or more hyphens (dashes)
+    docker_tag_name(
+        &metadata
+            .workspace_root
+            .file_name()
+            .expect("workspace_root can't end in `..`")
+            .to_string_lossy(),
+    )
+}
+
+fn docker_tag_name(file_name: &str) -> String {
+    // a valid image name consists of the following:
+    // - lowercase ASCII letters
+    // - digits
+    // - a period
+    // - 1-2 underscores
+    // - 1 or more hyphens (dashes)
+    let mut result = String::new();
+    let mut consecutive_underscores = 0;
```

Comment:
This needs to track state to ensure we don't have more than 2 consecutive underscores, or else this would be a lot easier/idiomatic.

Reply:
why not just remove any consecutive underscore?

```rs
name.to_ascii_lowercase().replace("__", "_").chars().filter(|c| c.is_ascii_digit() | matches!(c, '-' | '_'))
```

Reply:
reverse the order of the replace? so raplce after filtering

Reply:
I still don't think that works, since it doesn't match a pattern. We'd need a regex. For example:

```rust
# produces `package___test_name`, which is invalid
# we want `package_test_name` or `package__test_name`.
docker_tag_name("pAcKaGe_____test_name")  
```

Sine we can have 6, 7, or more ones, we'd have to do an iterative replace, which sounds like a bad idea.